extensions-core/orc-extensions/src/main/java/org/apache/druid/data/input/orc/OrcStructConverter.java

```diff
@@ -0,0 +1,239 @@
+        boolean     org.apache.hadoop.io.BooleanWritable
+        char        org.apache.hadoop.io.Text
+        date        org.apache.hadoop.hive.serde2.io.DateWritable
+        decimal     org.apache.hadoop.hive.serde2.io.HiveDecimalWritable
+        double      org.apache.hadoop.io.DoubleWritable
+        float       org.apache.hadoop.io.FloatWritable
+        int         org.apache.hadoop.io.IntWritable
+        smallint    org.apache.hadoop.io.ShortWritable
+        string      org.apache.hadoop.io.Text
+        timestamp   org.apache.orc.mapred.OrcTimestamp
+        tinyint     org.apache.hadoop.io.ByteWritable
+        varchar     org.apache.hadoop.io.Text
+     */
+    switch (fieldDescription.getCategory()) {
+      case STRING:
+      case CHAR:
+      case VARCHAR:
+        return ((Text) field).toString();
+      case BOOLEAN:
+        return ((BooleanWritable) field).get();
+      case BYTE:
+        return ((ByteWritable) field).get();
+      case SHORT:
+        return ((ShortWritable) field).get();
+      case INT:
+        return ((IntWritable) field).get();
+      case LONG:
+        return ((LongWritable) field).get();
+      case FLOAT:
+        return ((FloatWritable) field).get();
+      case DOUBLE:
+        return ((DoubleWritable) field).get();
+      case DECIMAL:
+        return ((HiveDecimalWritable) field).getHiveDecimal().doubleValue();
+      case TIMESTAMP:
+        return ((OrcTimestamp) field).getTime();
+      case DATE:
+        return DateTimes.utc(((DateWritable) field).get().getTime());
+      case BINARY:
+        byte[] bytes = ((BytesWritable) field).getBytes();
+        if (binaryAsString) {
+          return StringUtils.fromUtf8(bytes);
+        } else {
+          return bytes;
+        }
+      default:
+        return null;
+    }
+  }
+
+  private boolean binaryAsString;
+  private Object2IntMap<String> fieldIndexCache;
+
+  OrcStructConverter(boolean binaryAsString)
+  {
+    this.binaryAsString = binaryAsString;
+  }
+
+  /**
+   * Convert a orc struct field of the "root" {@link OrcStruct} that represents the "row". This method has a cache of
+   * field names to field index that is ONLY valid for this {@link OrcStruct}, and should not be used for
+   * nested {@link OrcStruct} fields of the row. Looks up field index by field name, and delegates to
+   * {@link OrcStructConverter#convertField(OrcStruct, int)}.
+   */
+  @Nullable
+  Object convertRootField(OrcStruct struct, String fieldName)
+  {
+    // this cache is only valid for the root level, to skip the indexOf on fieldNames to get the fieldIndex.
+    TypeDescription schema = struct.getSchema();
+    final List<String> fields = schema.getFieldNames();
+    if (fieldIndexCache == null) {
+      fieldIndexCache = new Object2IntOpenHashMap<>(fields.size());
+      for (int i = 0; i < fields.size(); i++) {
+        fieldIndexCache.put(fields.get(i), i);
+      }
+    }
+    WritableComparable wc = struct.getFieldValue(fieldName);
```

Comment:
Variable is not used. I assume that calling `struct.getFieldValue()` doesn't have desirable side effects. I'll delete this line.

Reply:
Deleted here: #7738

Reply:
Ouch, yeah this is not supposed to be there, I missed it and worse, it is sort of defeating the purpose of the field index cache since it's still causing `indexOf` to get called.. 

Thanks for fixing this, I'm going to open a PR to 0.15 branch to effectively backport this part of #7738 because this is a performance issue for data with lots of fields.